.github/workflows/main.yml

```
@@ -27,9 +27,23 @@ jobs:
         run: |
           pip install --upgrade pip
           pip install codecov tox tox-gh-actions
+          sudo apt-get update
+          sudo apt-get install --yes --no-install-recommends -V bubblewrap
 
       - name: Run tests with tox
         run: tox
 
+      - name: Run offline tests with tox without access to the Internet
+        run: |
+          # We're using bwrap/bubblewrap to take Internet access
+          # away from tox so that we'll notice whenever some new test
+          # is missing @pytest.mark.online decoration
+          bwrap \
+            --dev-bind / / \
+            --die-with-parent \
+            --unshare-net \
```

```suggestion
         unshare --map-root-user --net \
```

`unshare` is in util-linux, so should be installed by default.

---

@jspricke interesting idea but for some reason even offline tests fail wehn called this way. E.g. `unshare --map-root-user --net -- tox -e py310-requests-urllib3-1 -- -m 'not online' tests/integration/test_basic.py` with the fixed `runtests.sh` will show five failing tests:

```
FAILED tests/integration/test_basic.py::test_nonexistent_directory - OSError: [Errno 101] Network is unreachable
FAILED tests/integration/test_basic.py::test_unpatch - OSError: [Errno 101] Network is unreachable
FAILED tests/integration/test_basic.py::test_basic_json_use - OSError: [Errno 101] Network is unreachable
FAILED tests/integration/test_basic.py::test_patched_content - OSError: [Errno 101] Network is unreachable
FAILED tests/integration/test_basic.py::test_patched_content_json - OSError: [Errno 101] Network is unreachable
```

So if we switch to `unshare`, we'll need to call it some other way. Any ideas?

---

Can you try:

`unshare --map-root-user --net -- sh -c 'ip link set lo up; tox...`

---

@jspricke cool! Worked well locally, let's see how the CI likes it…